src/Storage/OWSIncomingMessageFinder.m

```diff
@@ -0,0 +1,156 @@
+
+NSString *const OWSIncomingMessageFinderColumnTimestamp = @"OWSIncomingMessageFinderColumnTimestamp";
+NSString *const OWSIncomingMessageFinderColumnSourceId = @"OWSIncomingMessageFinderColumnSourceId";
+NSString *const OWSIncomingMessageFinderColumnSourceDeviceId = @"OWSIncomingMessageFinderColumnSourceDeviceId";
+
+@interface OWSIncomingMessageFinder ()
+
+@property (nonatomic, readonly) YapDatabase *database;
+@property (nonatomic, readonly) YapDatabaseConnection *dbConnection;
+@property (nonatomic, readonly) BOOL isExtensionRegistered;
+
+@end
+
+@implementation OWSIncomingMessageFinder
+
+@synthesize dbConnection = _dbConnection;
+
+#pragma mark - init
+
+- (instancetype)init
+{
+    OWSAssert([TSStorageManager sharedManager].database != nil);
+
+    return [self initWithDatabase:[TSStorageManager sharedManager].database];
+}
+
+- (instancetype)initWithDatabase:(YapDatabase *)database
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    _database = database;
+
+    return self;
+}
+
+#pragma mark - properties
+
+- (YapDatabaseConnection *)dbConnection
+{
+    @synchronized (self) {
+        if (!_dbConnection) {
+            _dbConnection = self.database.newConnection;
+        }
+    }
+    return _dbConnection;
+}
+
+#pragma mark - YAP integration
+
+- (YapDatabaseSecondaryIndex *)indexExtension
+{
+    YapDatabaseSecondaryIndexSetup *setup = [YapDatabaseSecondaryIndexSetup new];
+
+    [setup addColumn:OWSIncomingMessageFinderColumnTimestamp withType:YapDatabaseSecondaryIndexTypeInteger];
+    [setup addColumn:OWSIncomingMessageFinderColumnSourceId withType:YapDatabaseSecondaryIndexTypeText];
+    [setup addColumn:OWSIncomingMessageFinderColumnSourceDeviceId withType:YapDatabaseSecondaryIndexTypeInteger];
+
+    YapDatabaseSecondaryIndexWithObjectBlock block = ^(YapDatabaseReadTransaction *transaction,
+        NSMutableDictionary *dict,
+        NSString *collection,
+        NSString *key,
+        id object) {
+        if ([object isKindOfClass:[TSIncomingMessage class]]) {
+            TSIncomingMessage *incomingMessage = (TSIncomingMessage *)object;
+
+            [dict setObject:@(incomingMessage.timestamp) forKey:OWSIncomingMessageFinderColumnTimestamp];
+            [dict setObject:incomingMessage.authorId forKey:OWSIncomingMessageFinderColumnSourceId];
+            [dict setObject:@(incomingMessage.sourceDeviceId) forKey:OWSIncomingMessageFinderColumnSourceDeviceId];
+        }
+    };
+
+    YapDatabaseSecondaryIndexHandler *handler = [YapDatabaseSecondaryIndexHandler withObjectBlock:block];
+
+    return [[YapDatabaseSecondaryIndex alloc] initWithSetup:setup handler:handler];
+}
+
+- (void)asyncRegisterExtension
+{
+    DDLogInfo(@"%@ registering async.", self.tag);
+    [self.database registerExtension:self.indexExtension withName:OWSIncomingMessageFinderExtensionName];
+}
+
+- (void)registerExtension
+{
+    OWSAssert(NO);
```

Comment:
It's explained in the subsequent log message, which was originally a comment, but I figured it was more useful as a log message and silly to copy/paste it as a comment. Would you have read it if the log message came before the assert?